Introduce to first stable version

CarlosAMS123 opened the pull request:
- First stable version -Performance
-No duplication
-Clean
-New SDK

Commit:
Introduce to first stable version

- First stable version
-Performance
-No duplication
-Clean
-New SDK

yamil-undostres commented on .vscode/launch.json:
```diff
@@ -0,0 +1,14 @@
+{
```
Añadir el directorio .vscode al .gitignore

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+        $this->supports = ['products', 'refunds', 'add_payment_method'];
+        $this->icon = WC_UDT_PLUGIN_URL . '/assets/images/icon.png';
+        /** INITS **/
+        self::init_form_fields();
```
init_form_fields no es estática pero está llamada estáticamente.

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+     */
+    public function process_order($order_id, $status): array
+    {
+        $order = self::get_order($order_id);
```
Mismo aquí, get_order no es estática

CarlosAMS123 replied:
Done.

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+    {
+        $order = self::get_order($order_id);
+        if ($order === null) return ['code' => 404, 'message' => 'Orden no encontrada.'];
+        if (!self::is_udt_order($order)) return ['code' => 500, 'message' => 'Orden no creada por UnDosTres.'];
```
Igual

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+     *
+     * Redirect to create api, and then it's intercepted by hook to redirect to UnDosTres.
+     */
+    public function process_payment($order_id): array
```
¿Esta función se usa en algún lado?

CarlosAMS123 replied:
Simon

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    /** FORMATTING OF PRODUCTS LIST **/
+    private function get_products($order_items): array
```
Sugerencia: renombrar a algo más indicativo de lo que hace, e.g. format_products_for_udt

CarlosAMS123 replied:
Idk, la funcion trae productos y esta dentro de la clase udt

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+        ];
+        $response = SASDK::createPayment($request);
+        self::log(sprintf("%s -> Payment url request send: %s \nReceive:\n%s", __METHOD__, json_encode($request), json_encode($response)));
+        if ($response["code"] !== 200) {
```
Este código proviene de los códigos de las excepciones de PHP. ¿Estamos seguros que esas excepciones nunca lanzan código 200?

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    /** TAX GETTER **/
+    private function get_total_taxes($taxes): int
```
Esta función se puede reducir a una línea mediante la función array_sum

yamil-undostres replied:
Se puede reducir esta función a una o dos líneas mediante un array_reduce.

yamil-undostres commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+            }
+            $response = SASDK::cancelOrder((string)$order_id);
+            self::log(sprintf("%s -> Admin order cancel, order: %s, received: %s", __METHOD__, $order_id, json_encode($response)));
+            if ($response["code"] !== 200)
```
Si aquí ocurre que la cancelación falló (código != 200) y el estado anterior era 'processing', el status de la orden se queda en 'refunded' en vez de regresar a 'processing'

CarlosAMS123 replied:
Excelente anotacion

yamil-undostres commented on woocommerce-gateway-undostres.php:
```diff
@@ -0,0 +1,153 @@
+            {
+            }
+
+            /** OVERRIDE WAKEUP METHOD TO PREVENT UNSERIALIZE **/
```
En este caso tal vez también se quiera impedir que el objeto se serialice en primer lugar al sobreescribir __sleep()

yamil-undostres commented on woocommerce-gateway-undostres.php:
```diff
@@ -0,0 +1,153 @@
+                require_once(WC_UDT_PLUGIN_PATH . '/includes/endpoints/class-wc-undostres-callback.php');
+                require_once(WC_UDT_PLUGIN_PATH . '/includes/endpoints/class-wc-undostres-redirect.php');
+                require_once(WC_UDT_PLUGIN_PATH . '/includes/endpoints/class-wc-undostres-status.php');
+                $callback = new WC_REST_UDT_CALLBACK_CONTROLLER($this->get_gateway());
```
Sugerencia: esta función se podría encoger y optimizar (aunque solo un poco) si las funciones register_routes fueran estáticas

yamil-undostres commented on includes/utility/class-wc-undostres-logger.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public static $logger;
+
+    public static function log($message, bool $encode = false)
```
Según veo el parámetro $encode nunca se utiliza. ¿Se puede quitar?

yamil-undostres committed:
fix: Spelling and grammar in admin messages

yamil-undostres committed:
fix: Spelling and grammar in README

Commit:
Pul request fixes

-$this en vez de self
-Error en cancelacion de orden con procesing y refunded
-Limpieza
-Sleep override para que no se serielize

Commit:
Merge branch 'integration' of https://github.com/undostres-com-mx/woocommerce-gateway-undostres into integration

VicLimaAlamo commented on README.md:
```diff
@@ -79,18 +81,18 @@ send_timeout                600;
 
-- PHP Debug, on VsCode.
-- Xdebug helper, on web browser.
+- PHP Debug, in VsCode.
+- Xdebug helper, in the web browser.
```
"on" is a better eng there d: @CarlosAMS123

CarlosAMS123 replied:
@VicLimaAlamo excelent, on it

VicLimaAlamo commented on assets/css/styles.css:
```diff
@@ -0,0 +1,10 @@
+kueskipay-widget,
+.nonoplazo_wsppc_div_block,
+aplazo-placement,
```
no falta aca tmb mercado pago y conekta? @CarlosAMS123 @javierhc25

CarlosAMS123 replied:
@VicLimaAlamo
Los otros métodos de pago son desactivados con hook, esto solo es para para eliminar publicidad intrusiva que forzan los otros plugins.

VicLimaAlamo replied:
ok, no hemos detectado entonces publicidad de los otros?

VicLimaAlamo commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+     *
+     * Redirect to create api, and then it's intercepted by hook to redirect to UnDosTres.
+     */
+    public function process_payment($order_id): array
```
woocommerce siempre trabaja con php > 5? sino esta funcion puede romperlo

CarlosAMS123 replied:
@VicLimaAlamo 
Wordpress acepta php 5.6.20 + (Recomienda 7.4)
Sin embargo, Woocommerce desde la 6.5 (Que justo es la minima version con la que somos compatibles) requiere de minimo php 7.2

VicLimaAlamo replied:
pero todos tienen esa version?
deberiamos decir que somos entonces compatibles para esa version o cambiar el codigo si queremos ser compatibles para los otros...

VicLimaAlamo commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+            $this->log(sprintf("%s -> Refund request, order: %s, amount: %s, received: %s", __METHOD__, $order_id, $amount, json_encode($response)));
+            if ($response["code"] !== 200)
+                $order->add_order_note('El reembolso no se ha podido procesar: ' . $response["status"]);
+            return $response["code"] === 200;
```
no debemos devolver aca otra cosa? no 200 pq justo arriba esta el caso de fallo, esto deberia ser error 400 no? @CarlosAMS123 @javierhc25

CarlosAMS123 replied:
@VicLimaAlamo lo que regresa es true o false, ya que process_refund es una implementacion de WC_Payment_Gateway, si se regresa true significa que el reembolso fue correcto, si se regresa false que hubo un error asi que no se cambian status ni nada
Y la parte de arriba es para añadir una nota en frontend de el porque del error en el rembolso.

VicLimaAlamo replied:
ok

VicLimaAlamo commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    /** FORMATTING OF PRODUCTS LIST **/
+    private function get_products($order_items): array
```
woocommerce siempre trabaja con php > 5? sino esta funcion puede romperlo

VicLimaAlamo commented on includes/class-wc-gateway-undostres.php:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    /** TAX GETTER **/
+    private function get_total_taxes($taxes): int
```
woocommerce siempre trabaja con php > 5? sino esta funcion puede romperlo

VicLimaAlamo commented on includes/endpoints/class-wc-undostres-callback.php:
```diff
@@ -0,0 +1,47 @@
+        );
+    }
+
+    public function callback($data): WP_REST_Response
```
woocommerce siempre trabaja con php > 5? sino esta funcion puede romperlo

VicLimaAlamo commented on includes/endpoints/class-wc-undostres-callback.php:
```diff
@@ -0,0 +1,47 @@
+            $response = $this->gateway->process_order($orderId, $status);
+            return new WP_REST_Response($response);
+        } catch (Exception $e) {
+            return new WP_REST_Response(['code' => 500, 'message' => $e->getMessage()]);
```
solo estamos mandando esto a ellos, pero para nosotros no guardamos algo? @CarlosAMS123 @javierhc25

CarlosAMS123 replied:
@VicLimaAlamo 
A que te refieres? Si es en cuestión de log en la línea 40 se guarda un log del cambio si se requiere algún log más se podría añadir

VicLimaAlamo replied:
si, mi punto es que tu codigo a veces no llegara a la linea 40, solo a la 38 o a la 39, de ahi se salta al catch y no hay log del fallo

CarlosAMS123 replied:
Añadir log de error

VicLimaAlamo commented on includes/endpoints/class-wc-undostres-redirect.php:
```diff
@@ -0,0 +1,39 @@
+    {
+        $orderId = $data->get_param('orderId');
+        $this->gateway->log(sprintf("%s -> Redirect de la orden: %s", __METHOD__, $orderId ?? 'NULL'));
+        if (wp_redirect($this->gateway->get_redirect_url($orderId))) exit;
```
yo creo que hay que poner al menos un comentario aca, cualquier linea debajo de esto es probable que no se ejecute @CarlosAMS123 @javierhc25

VicLimaAlamo commented on includes/endpoints/class-wc-undostres-status.php:
```diff
@@ -0,0 +1,49 @@
+            else $response = ['code' => 200, 'message' => 'Ok.', 'status' => $order->get_status()];
+            return new WP_REST_Response($response);
+        } catch (Exception $e) {
+            return new WP_REST_Response(['code' => 500, 'message' => $e->getMessage()]);
```
solo estamos mandando esto a ellos, pero para nosotros no guardamos algo? @CarlosAMS123 @javierhc25
